File: CSharp/src/CnpjCalculator.cs
using System.Text.RegularExpressions;

namespace CnpjValidator;

public static class CnpjCalculator
{
    private static readonly Regex OnlyBaseAlfa = new Regex(@"^[0-9A-Z]{12}$", RegexOptions.Compiled);
    private static readonly Regex OnlyBaseNum = new Regex(@"^[0-9]{12}$", RegexOptions.Compiled);

    /// <summary>
    /// Método único para validar CNPJ (clássico ou alfanumérico).
    /// </summary>
    public static bool ValidarCNPJ(string cnpj)
    {
        if (string.IsNullOrWhiteSpace(cnpj)) return false;

        string limpo = LimparMascara(cnpj).ToUpperInvariant();

        if (limpo.Length != 14) return false;

        string base12 = limpo.Substring(0, 12);
        string dvStr = limpo.Substring(12, 2);

        // se base12 é só dígitos => clássico
        if (OnlyBaseNum.IsMatch(base12))
        {
            return CalcularDvNumerico(base12, dvStr);
        }
        // se base12 contém letras => alfanumérico
        else if (OnlyBaseAlfa.IsMatch(base12))
        {
            return CalcularDvAlfanumerico(base12, dvStr);
        }

        return false;
    }

    /// <summary>
    /// Valida os dígitos verificadores de um CNPJ alfanumérico.
    /// </summary>
    private static bool CalcularDvAlfanumerico(string base12, string dvInformado)
    {
        if (string.IsNullOrWhiteSpace(base12) || string.IsNullOrWhiteSpace(dvInformado)) return false;
        base12 = LimparMascara(base12).ToUpperInvariant();
        dvInformado = dvInformado.Trim();

        if (!OnlyBaseAlfa.IsMatch(base12) || !Regex.IsMatch(dvInformado, @"^\d{2}$")) return false;

        int d1 = CalcularDv(base12, false);
        int d2 = CalcularDv(base12 + d1, true);
        return dvInformado == $"{d1}{d2}";
    }

    /// <summary>
    /// Valida os dígitos verificadores de um CNPJ tradicional (numérico).
    /// </summary>
    private static bool CalcularDvNumerico(string base12Numerica, string dvInformado)
    {
        if (string.IsNullOrWhiteSpace(base12Numerica) || string.IsNullOrWhiteSpace(dvInformado)) return false;
        base12Numerica = LimparMascara(base12Numerica);
        dvInformado = dvInformado.Trim();

        if (!OnlyBaseNum.IsMatch(base12Numerica) || !Regex.IsMatch(dvInformado, @"^\d{2}$")) return false;

        int d1 = CalcularDv(base12Numerica, false);
        int d2 = CalcularDv(base12Numerica + d1, true);
        return dvInformado == $"{d1}{d2}";
    }

    private static int CalcularDv(string texto, bool tamanhoComDv)
    {
        int len = tamanhoComDv ? 13 : 12;
        if (texto.Length != len)
            throw new ArgumentException($"Entrada deve ter {len} caracteres.", nameof(texto));

        int[] pesos = GerarPesos(len);

        int soma = 0;
        for (int i = 0; i < len; i++)
        {
            char c = texto[i];
            int valor = ValorParaCalculo(c);
            soma += valor * pesos[i];
        }

        int resto = soma % 11;
        return (resto == 0 || resto == 1) ? 0 : 11 - resto;
    }

    private static int ValorParaCalculo(char c)
    {
        if (char.IsDigit(c)) return c - '0';
        char u = char.ToUpperInvariant(c);
        if (u >= 'A' && u <= 'Z') return u - 48;
        throw new ArgumentException($"Caractere inválido: '{c}'. Use apenas [A-Z0-9].");
    }

    private static int[] GerarPesos(int length)
    {
        var pesos = new int[length];
        int w = 2;
        for (int i = length - 1; i >= 0; i--)
        {
            pesos[i] = w;
            w++;
            if (w > 9) w = 2;
        }
        return pesos;
    }

    private static string LimparMascara(string s)
        => new string((s ?? "").Where(ch => ch != '.' && ch != '/' && ch != '-' && !char.IsWhiteSpace(ch)).ToArray());
}

File: Python/CnpjCalculator.py
import re

class CnpjCalculator:
    @staticmethod
    def validar_cnpj(cnpj: str) -> bool:
        limpo = re.sub(r'[^A-Z0-9]', '', cnpj.upper())
        if len(limpo) != 14:
            return False

        base12, dv = limpo[:12], limpo[12:]

        if re.fullmatch(r"\d{12}", base12):
            return CnpjCalculator._calcular_dv_numerico(base12, dv)
        elif re.fullmatch(r"[0-9A-Z]{12}", base12):
            return CnpjCalcul._calcular_dv_alfanumerico(base12, dv)
        return False

    @staticmethod
    def _calcular_dv_numerico(base12: str, dv: str) -> bool:
        if not re.fullmatch(r"\d{12}", base12) or not re.fullmatch(r"\d{2}", dv):
            return False
        d1 = CnpjCalculator._calcular_um_dv(base12, False)
        d2 = CnpjCalculator._calcular_um_dv(base12 + str(d1), True)
        return dv == f"{d1}{d2}"

    @staticmethod
    def _calcular_dv_alfanumerico(base12: str, dv: str) -> bool:
        if not re.fullmatch(r"[0-9A-Z]{12}", base12) or not re.fullmatch(r"\d{2}", dv):
            return False
        d1 = CnpjCalculator._calcular_um_dv(base12, False)
        d2 = CnpjCalculator._calcular_um_dv(base12 + str(d1), True)
        return dv == f"{d1}{d2}"

    @staticmethod
    def _calcular_um_dv(texto: str, com_dv: bool) -> int:
        length = 13 if com_dv else 12
        pesos = CnpjCalculator._gerar_pesos(length)
        soma = sum(CnpjCalculator._valor_char(texto[i]) * pesos[i] for i in range(length))
        resto = soma % 11
        return 0 if resto in (0, 1) else 11 - resto

    @staticmethod
    def _valor_char(c: str) -> int:
        o = ord(c)
        if (48 <= o <= 57) or (65 <= o <= 90):
            return o - 48
        raise ValueError(f"Caractere inválido: {c}")

    @staticmethod
    def _gerar_pesos(length: int) -> list[int]:
        pesos = [0] * length
        w = 2
        for i in range(length - 1, -1, -1):
            pesos[i] = w
            w += 1
            if w > 9:
                w = 2
        return pesos




